functions/index.js

```diff
@@ -43,5 +44,118 @@ const onAnswerMigrateToUserCollection = functions
+        if (deviceGroupResponse.failureCount > 0) {
+          await runFCMOperation('remove', notificationKeyName, notificationKey, deviceGroupResponse.failedRegistrationTokens)
+        }
+      } catch (error) {
+        console.log(error)
+      }
+    }
+  })
+
+const fetchNotificationRequest = async (url, method, body) => {
+  try {
+    const options = {
+      method: method,
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `key = ${functions.config().fcm.api_key}`,
+        project_id: functions.config().fcm.sender_id
+      }
+    }
+    if (method === 'post') {
+      options.body = JSON.stringify(body)
+    }
+    const response = await fetch(url, options)
+    return await response.text()
+  } catch (error) {
+    console.log(error)
+    return null
+  }
+}
+
+const getNotificationKey = async (notificationKeyName) => {
+  const url = `https://fcm.googleapis.com/fcm/notification?notification_key_name=${notificationKeyName}`
+  try {
+    const result = await fetchNotificationRequest(url, 'get', {})
+    const json = JSON.parse(result)
+    if (json.notification_key !== undefined) {
+      return json.notification_key
+    }
+  } catch (error) {
+    console.log(error)
+  }
+  return null
+}
+
+const createNotificationGroup = async (notificationKeyName, fcmToken) => {
+  const url = 'https://fcm.googleapis.com/fcm/notification'
+  const data = {
+    operation: 'create',
+    notification_key_name: notificationKeyName,
+    registration_ids: [fcmToken]
+  }
+  await fetchNotificationRequest(url, 'post', data)
+}
+
+const runFCMOperation = async (operation, notificationKeyName, notificationKey, fcmTokens) => {
+  const url = 'https://fcm.googleapis.com/fcm/notification'
+  const body = {
+    operation: operation,
+    notification_key_name: notificationKeyName,
+    notification_key: notificationKey,
+    registration_ids: fcmTokens
+  }
+  await fetchNotificationRequest(url, 'post', body)
+}
+
+const storeFCMTokenInUserGroup = functions
+  .https
+  .onCall(async (data, context) => {
+    const fcmToken = data.fcmToken
+    const uid = context.auth.uid
+    const notificationKeyName = 'appUser-' + uid
+    const notificationKey = await getNotificationKey(notificationKeyName)
+
+    if (notificationKey === null) {
+      await createNotificationGroup(notificationKeyName, fcmToken)
+    } else {
+      await runFCMOperation('add', notificationKeyName, notificationKey, [fcmToken])
+    }
+  })
+
+const removeFCMTokenInUserGroup = functions
+  .https
+  .onCall(async (data, context) => {
+    const fcmToken = data.fcmToken
+    const uid = context.auth.uid
+    const notificationKeyName = 'appUser-' + uid
+    const notificationKey = await getNotificationKey(notificationKeyName)
+
+    if (notificationKey !== null) {
+      await runFCMOperation('remove', notificationKeyName, notificationKey, [fcmToken])
+    }
+  })
```

Comment:
i'm having trouble seeing the execution flow of this. On a https call, those two functions will fire (in "parallel"). 
On `notificationKey !== null`:

-  `storeFCMTokenInUserGroup` will do `runFCMOperation('add')`, and 

-  `removeFCMTokenInUserGroup` will do `runFCMOperation('remove')`.

Can you elaborate a little bit the expected behaviour (in this discussion)?

Reply:
This functions work a little bit different to the ones that listens to database changes. 
This functions are callables from the client by the function's name.

From MessagingService module we can call them like this...

-`
const removeFCMTokenInUserGroup = firebase.functions().httpsCallable('removeFCMTokenInUserGroup')`
`await removeFCMTokenInUserGroup({ fcmToken: fcmToken })`

-`const storeFCMTokenInUserGroup = firebase.functions().httpsCallable('storeFCMTokenInUserGroup')`
`await storeFCMTokenInUserGroup({ fcmToken: fcmToken })`



- `storeFCMTokenInUserGroup` is called when users **login**, it's purpose is to save the fcm token into a existing user's devices group, or into a new user's devices group if there isn't one.

- `removeFCMTokenInUserGroup` is called when users **logout**, it's purpose is to remove the fcm token in a existing user's devices group. FCM api automatically deletes empty groups that's why we need to create them again when the user login.

Reply:
Amazing! I shall write this as a recipe.